bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,437 @@
+
+	if upgradedMachines == len(ownedMachines) {
+		if err := r.markUpgradeDone(ctx, scope); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to mark upgrade as done: %w", err)
+		}
+
+		log.Info("All machines are upgraded")
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+}
+
+// markUpgradeInProgress marks the MachineDeployment as in-place upgrade in-progress.
+func (r *MachineDeploymentReconciler) markUpgradeInProgress(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	ant := scope.MachineDeployment.Annotations
+	if ant == nil {
+		ant = make(map[string]string)
+	}
+
+	// clean up
+	delete(ant, bootstrapv1.InPlaceUpgradeReleaseAnnotation)
+
+	ant[bootstrapv1.InPlaceUpgradeToAnnotation] = scope.UpgradeTo
+	ant[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeInProgressStatus
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeInProgressEvent,
+		"MachineDeployment %s marked as in-place upgrade in-progress",
+		scope.MachineDeployment.Name,
+	)
+	return nil
+}
+
+// markUpgradeDone marks the MachineDeployment as in-place upgrade done.
+func (r *MachineDeploymentReconciler) markUpgradeDone(ctx context.Context, scope *MachineDeploymentUpgradeScope) error {
+	ant := scope.MachineDeployment.Annotations
+	if ant == nil {
+		ant = make(map[string]string)
+	}
+
+	// clean up
+	delete(ant, bootstrapv1.InPlaceUpgradeToAnnotation)
+
+	ant[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeDoneStatus
+	ant[bootstrapv1.InPlaceUpgradeReleaseAnnotation] = scope.UpgradeTo
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeNormal,
+		bootstrapv1.InPlaceUpgradeDoneEvent,
+		"MachineDeployment %s marked as in-place upgrade done",
+		scope.MachineDeployment.Name,
+	)
+	return nil
+}
+
+// markUpgradeFailed marks the MachineDeployment as in-place upgrade failed.
+func (r *MachineDeploymentReconciler) markUpgradeFailed(ctx context.Context, scope *MachineDeploymentUpgradeScope, failedM *clusterv1.Machine) error {
+	ant := scope.MachineDeployment.Annotations
+	if ant == nil {
+		ant = make(map[string]string)
+	}
+
+	// clean up
+	delete(ant, bootstrapv1.InPlaceUpgradeReleaseAnnotation)
+
+	ant[bootstrapv1.InPlaceUpgradeStatusAnnotation] = bootstrapv1.InPlaceUpgradeFailedStatus
+
+	if err := scope.PatchHelper.Patch(ctx, scope.MachineDeployment); err != nil {
+		return fmt.Errorf("failed to patch: %w", err)
+	}
+
+	r.recorder.Eventf(
+		scope.MachineDeployment,
+		corev1.EventTypeWarning,
+		bootstrapv1.InPlaceUpgradeFailedEvent,
+		"MachineDeployment %q marked as in-place upgrade failed. Upgrade failed for machine %q",
```

Comment:
We could include the error message here. Would this make sense?

Reply:
This is a really good idea, but I think it's not easily feasible right now. The single machine in-place upgrade reconciler (which has access to the original error) publishes an event for that but does not include it anywhere else: https://github.com/canonical/cluster-api-k8s/blob/main/bootstrap/controllers/upgrade_controller.go#L189-L204